met-web/src/components/survey/edit/ActionContext.tsx

```diff
@@ -0,0 +1,169 @@
+import React, { createContext, useState, useEffect } from 'react';
+import { useAppDispatch } from 'hooks';
+import { useNavigate, useParams } from 'react-router-dom';
+import { openNotification } from 'services/notificationService/notificationSlice';
+import { getEmailVerification } from 'services/emailVerificationService';
+import { getEngagement } from 'services/engagementService';
+import { getSubmissionByToken, updateSubmission } from 'services/submissionService';
+import { Engagement } from 'models/engagement';
+import { PublicSubmission } from 'models/surveySubmission';
+
+type EditSurveyParams = {
+    token: string;
+    engagementId: string;
+};
+
+interface EditSurveyContext {
+    token?: string;
+    isTokenValid: boolean;
+    handleSubmit: () => void;
+    isSubmitting: boolean;
+    savedEngagement: Engagement | null;
+    isLoading: boolean;
+    loadEngagement: null | (() => void);
+    submission: PublicSubmission | null;
+}
+
+export const ActionContext = createContext<EditSurveyContext>({
+    isTokenValid: true,
+    handleSubmit: () => {
+        return;
+    },
+    isSubmitting: false,
+    savedEngagement: null,
+    isLoading: true,
+    loadEngagement: null,
+    submission: null,
+});
+
+export const ActionProvider = ({ children }: { children: JSX.Element }) => {
+    const navigate = useNavigate();
+    const dispatch = useAppDispatch();
+    const { engagementId, token } = useParams<EditSurveyParams>();
+    const [isTokenValid, setTokenValid] = useState(true);
+    const [isSubmitting, setIsSubmitting] = useState(false);
+    const [savedEngagement, setSavedEngagement] = useState<Engagement | null>(null);
+    const [isLoading, setIsLoading] = useState(true);
+    const [submission, setSubmission] = useState<PublicSubmission | null>(null);
+
+    useEffect(() => {
+        fetchData();
+    }, []);
+
+    const fetchData = async () => {
+        await loadEngagement();
+        await verifyToken();
+        await loadSubmission();
+        setIsLoading(false);
+    };
+
+    const verifyToken = async () => {
+        if (!token) {
+            navigate(`/404`);
+            return;
+        }
+
+        try {
+            const verification = await getEmailVerification(token);
+            if (!verification) {
+                throw new Error('Invalid token');
+            }
+        } catch (error) {
+            dispatch(
+                openNotification({
+                    severity: 'error',
+                    text: 'Verification token is invalid.',
+                }),
+            );
+            setTokenValid(false);
+        }
+    };
+
+    const loadSubmission = async () => {
+        if (token && isTokenValid)
```

Comment:
I would consider returning early here rather than wrapping everything with the if. if(!token || !isTokenValid) return;. And rest of logic is independent of it.